components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkUtils.java

```diff
@@ -2414,5 +2421,81 @@ public static FederatedAssociationManager getFederatedAssociationManager() throw
         }
         return federatedAssociationManager;
     }
-}
 
+    /**
+     * Retrieves the unique user id of the given username. If the unique user id is not available, generate an id and
+     * update the userid claim in read/write userstores.
+     *
+     * @param tenantId    id of the tenant domain of the user
+     * @param userStoreDomain userstore of the user
+     * @param username        username
+     * @return unique user id of the user
+     * @throws UserSessionException
+     */
+    public static String resolveUserIdFromUsername(int tenantId, String userStoreDomain, String username) throws
+            UserSessionException {
+
+        try {
+            UserStoreManager userStoreManager = getUserStoreManager(tenantId, userStoreDomain);
+            try {
+                if (userStoreManager instanceof AbstractUserStoreManager) {
+                    String userId = ((AbstractUserStoreManager) userStoreManager).getUserIDFromUserName(username);
+
+                    // If only the user store is not read-only, update the user-id claim with the unique id.
+                    // Otherwise there will be no permission to update the userstore.
+                    if (StringUtils.isBlank(userId) && !userStoreManager.isReadOnly()) {
+                        userId = addUserId(username, userStoreManager);
+                    }
+                    return userId;
+                }
+                if (log.isDebugEnabled()) {
+                    log.debug("Provided user store manager for the user: " + username + ", is not an instance of the " +
+                            "AbstractUserStore manager");
+                }
+                throw new UserSessionException("Unable to get the unique id of the user: " + username + ".");
+            } catch (org.wso2.carbon.user.core.UserStoreException e) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Error occurred while retrieving Id for the user: " + username, e);
+                }
+                throw new UserSessionException("Unable to retrieve Id for the user: " + username);
+            }
+        } catch (UserStoreException e) {
+            throw new UserSessionException("Unable to retrieve Id for the user: " + username);
+        }
+    }
+
+    private static String addUserId(String username, UserStoreManager userStoreManager) {
+
+        String userId;
+        userId = UUIDGenerator.generateUUID();
```

Comment:
```suggestion
        userId = UUID.randomUUID().toString();
```

Reply:
This is the util method available in the user-core level. If we use the same UUID generate logic, we can customize the logic if needed, easily from one place

Reply:
This comes from the org.wso2.carbon.registry.core.utils package. There is no point to have a dependency to the registry from here.

Reply:
Changed